### PR TITLE
align cuda driver and runtime labels with the NFD kernel version label format

### PR DESCRIPTION
**Pull request by tariq1890**
Align the GFD cuda driver and runtime version labels with the kernel version labels of

 ```
      feature.node.kubernetes.io/kernel-version.full: 5.15.0-101-generic                                                                                                                       
      feature.node.kubernetes.io/kernel-version.major: "5"                                                                                                                              
      feature.node.kubernetes.io/kernel-version.minor: "15"                                                                                                                             
      feature.node.kubernetes.io/kernel-version.revision: "0" 
 ```

This adds the following labels:
```
nvidia.com/cuda.driver-version.major
nvidia.com/cuda.driver-version.minor
nvidia.com/cuda.driver-version.revision
nvidia.com/cuda.driver-version.full
nvidia.com/cuda.runtime-version.major
nvidia.com/cuda.runtime-version.minor
nvidia.com/cuda.runtime-version.full
```

Note that the existing labels:
```
nvidia.com/cuda.driver.major
nvidia.com/cuda.driver.minor
nvidia.com/cuda.driver.rev
nvidia.com/cuda.runtime.major
nvidia.com/cuda.runtime.minor
```
should be considered deprecated.

### CODE REVIEW

**Review by elezar — COMMENTED** [review]
Could we redo the labels to better align with what is done for the existing ones.

I mean the kernel ones.

**Comment by tariq1890** [comment]
@elezar  Done

**Review by ArangoGutierrez — APPROVED** [review]

File: internal/lm/nvml.go [hunk 1]
```diff
@@ -101,11 +101,21 @@ func newVersionLabeler(manager resource.Manager) (Labeler, error) {
 	}
 
 	labels := Labels{
+		// Deprecated labels
 		"nvidia.com/cuda.driver.major":  driverMajor,
 		"nvidia.com/cuda.driver.minor":  driverMinor,
 		"nvidia.com/cuda.driver.rev":    driverRev,
 		"nvidia.com/cuda.runtime.major": fmt.Sprintf("%d", *cudaMajor),
 		"nvidia.com/cuda.runtime.minor": fmt.Sprintf("%d", *cudaMinor),
```
Comment (elezar) [inline internal/lm/nvml.go:109]: My one question was whether we wanted the following set of labels instead:
```
		"nvidia.com/cuda.driver.major"
		"nvidia.com/cuda.driver.minor"
		"nvidia.com/cuda.driver.rev"
		"nvidia.com/cuda.driver.revision"
		"nvidia.com/cuda.driver.full"
		"nvidia.com/cuda.runtime.major"
		"nvidia.com/cuda.runtime.minor"
		"nvidia.com/cuda.runtime.revision"
		"nvidia.com/cuda.runtime.full"
```
So as to not have to deprecate labels except for `"nvidia.com/cuda.driver.rev"`?

@klueska do you have thoughts w.r.t the `-version` string inserted above?

File: internal/lm/nvml.go [hunk 2]
```diff
@@ -101,11 +101,21 @@ func newVersionLabeler(manager resource.Manager) (Labeler, error) {
...
+
+		// New labels
+		"nvidia.com/cuda.driver-version.major":    driverMajor,
+		"nvidia.com/cuda.driver-version.minor":    driverMinor,
+		"nvidia.com/cuda.driver-version.revision": driverRev,
+		"nvidia.com/cuda.driver-version.full":     driverVersion,
+		"nvidia.com/cuda.runtime-version.major":   fmt.Sprintf("%d", *cudaMajor),
+		"nvidia.com/cuda.runtime-version.minor":   fmt.Sprintf("%d", *cudaMinor),
+		"nvidia.com/cuda.runtime-version.full":    fmt.Sprintf("%d.%d", *cudaMajor, *cudaMinor),
```
Comment (elezar) [inline internal/lm/nvml.go:118]: Let's add an "always zero" `cuda.runtime-version.revision` aswell.

File: internal/lm/nvml.go [hunk 3]
```diff
@@ -101,11 +101,21 @@ func newVersionLabeler(manager resource.Manager) (Labeler, error) {
...
 		"nvidia.com/cuda.driver.minor":  driverMinor,
 		"nvidia.com/cuda.driver.rev":    driverRev,
 		"nvidia.com/cuda.runtime.major": fmt.Sprintf("%d", *cudaMajor),
 		"nvidia.com/cuda.runtime.minor": fmt.Sprintf("%d", *cudaMinor),
+
+		// New labels
+		"nvidia.com/cuda.driver-version.major":    driverMajor,
+		"nvidia.com/cuda.driver-version.minor":    driverMinor,
+		"nvidia.com/cuda.driver-version.revision": driverRev,
+		"nvidia.com/cuda.driver-version.full":     driverVersion,
```
Comment (elezar) [inline internal/lm/nvml.go:115]: Question: Do we want to generate a cannonical version string here?

**Review by elezar — COMMENTED** [review]
@tariq1890 thanks for this. Definitely a great addition. Some minor questions below.

File: internal/lm/nvml.go (hunk 3 above)
Comment (tariq1890) [reply internal/lm/nvml.go:115]: What is difference between the canonical version and the ful version string here?

File: internal/lm/nvml.go (hunk 2 above)
Comment (tariq1890) [reply internal/lm/nvml.go:118]: Got it. So do we know for certain that cuda versions will have a `patch` number in their semver ?

File: internal/lm/nvml.go (hunk 1 above)
Comment (tariq1890) [reply internal/lm/nvml.go:109]: I have a slight preference towards `driver-version` and `runtime-version` as it is more explicit.

File: internal/lm/nvml.go (hunk 3 above)
Comment (elezar) [reply internal/lm/nvml.go:115]: https://pkg.go.dev/golang.org/x/mod/semver#Canonical

A Canonical has the .MINOR and .PATCH (revision) included in the string -- that is to say it always consists of three components.

File: internal/lm/nvml.go (hunk 2 above)
Comment (elezar) [reply internal/lm/nvml.go:118]: I'm happy to not included it though, if you feel that this will be confusing.

File: internal/lm/nvml.go (hunk 3 above)
Comment (tariq1890) [reply internal/lm/nvml.go:115]: So If I understand correctly, aren't the full version and canonical versions the same thing here?

Do we get semver strings that include a build metadata component from the NVML API?

File: internal/lm/nvml.go (hunk 2 above)
Comment (tariq1890) [reply internal/lm/nvml.go:118]: If there is no #define declaration for the CUDA revision number, then I am not sure if there is any value adding an `always zero` label

Comment (elezar) [reply internal/lm/nvml.go:118]: Yes, let's not add the `.revision` label here then.

File: internal/lm/nvml.go (hunk 3 above)
Comment (elezar) [reply internal/lm/nvml.go:115]: No, we don't get that information from the API. We only get major, minor, and patch versions.

File: tests/expected-output-mig-mixed.txt [hunk 4]
```diff
@@ -4,6 +4,13 @@ nvidia\.com\/cuda\.driver\.minor=[0-9]+
 nvidia\.com\/cuda\.driver\.rev=[0-9]*
 nvidia\.com\/cuda\.runtime\.major=[0-9]+
 nvidia\.com\/cuda\.runtime\.minor=[0-9]+
+nvidia\.com\/cuda\.driver-version\.major=[0-9]+
+nvidia\.com\/cuda\.driver-version\.minor=[0-9]+
+nvidia\.com\/cuda\.driver-version\.revision=[0-9]*
+nvidia\.com\/cuda\.runtime-version\.major=[0-9]+
+nvidia\.com\/cuda\.runtime-version\.minor=[0-9]+
+nvidia\.com\/cuda\.driver-version\.full=([0-9]+)(.[0-9]+)(.[0-9]+)
```
Comment (elezar) [inline tests/expected-output-mig-mixed.txt:12]: nit: let's keep these sorted.

**Review by elezar — APPROVED** [review]
Thanks for the work here @tariq1890.

I'm good with this, pending @klueska weighing in on the `-version` string addition.

File: internal/lm/nvml.go (hunk 2 above)
Comment (tariq1890) [reply internal/lm/nvml.go:118]: Got it. Can we resolve this comment?

File: tests/expected-output-mig-mixed.txt (hunk 4 above)
Comment (tariq1890) [reply tests/expected-output-mig-mixed.txt:12]: For sure, will do

Comment (elezar) [reply tests/expected-output-mig-mixed.txt:12]: Sorr ... I just mean the ones you added.

Comment (tariq1890) [reply tests/expected-output-mig-mixed.txt:12]: Got it. Will update

Comment (tariq1890) [reply tests/expected-output-mig-mixed.txt]: Done

**Comment by elezar** [comment]
@tariq1890 I'm going to merge this. Could you create a follow-up that adds a config option to allow the deprecated labels to be disabled and optionally omit them.

Something along the lines of:
```
diff --git a/internal/lm/nvml.go b/internal/lm/nvml.go
index 7a4bcc4d..79dc2eb0 100644
--- a/internal/lm/nvml.go
+++ b/internal/lm/nvml.go
@@ -101,14 +101,6 @@ func newVersionLabeler(manager resource.Manager) (Labeler, error) {
 	}
 
 	labels := Labels{
-		// Deprecated labels
-		"nvidia.com/cuda.driver.major":  driverMajor,
-		"nvidia.com/cuda.driver.minor":  driverMinor,
-		"nvidia.com/cuda.driver.rev":    driverRev,
-		"nvidia.com/cuda.runtime.major": fmt.Sprintf("%d", *cudaMajor),
-		"nvidia.com/cuda.runtime.minor": fmt.Sprintf("%d", *cudaMinor),
-
-		// New labels
 		"nvidia.com/cuda.driver-version.major":    driverMajor,
 		"nvidia.com/cuda.driver-version.minor":    driverMinor,
 		"nvidia.com/cuda.driver-version.revision": driverRev,
@@ -117,7 +109,18 @@ func newVersionLabeler(manager resource.Manager) (Labeler, error) {
 		"nvidia.com/cuda.runtime-version.minor":   fmt.Sprintf("%d", *cudaMinor),
 		"nvidia.com/cuda.runtime-version.full":    fmt.Sprintf("%d.%d", *cudaMajor, *cudaMinor),
 	}
-	return labels, nil
+
+	if !config.useDeprecatedVersionLabels {
+		return labels, nil
+	}
+	deprecatedLabels := Labels{
+		"nvidia.com/cuda.driver.major":  driverMajor,
+		"nvidia.com/cuda.driver.minor":  driverMinor,
+		"nvidia.com/cuda.driver.rev":    driverRev,
+		"nvidia.com/cuda.runtime.major": fmt.Sprintf("%d", *cudaMajor),
+		"nvidia.com/cuda.runtime.minor": fmt.Sprintf("%d", *cudaMinor),
+	}
+	return Merge(deprecatedLabels, labels), nil
 }
 
 // newMigCapabilityLabeler creates a new MIG capability labeler using the provided NVML library.
```

although the exact mechanics could be changed.

**Pull request merged**